teps/0136-capture-traces-for-task-pod-events.md

```diff
@@ -0,0 +1,361 @@
+creation-date: '2023-06-14'
+last-updated: '2023-06-14'
+authors:
+- '@kmjayadeep'
+collaborators: []
+---
+
+# TEP-0136: Capture traces for task pod events
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+  - [Capturing step status update](#capturing-step-status-update)
+  - [Capturing pod failure event](#capturing-pod-failure-event)
+  - [POC](#poc)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [Conformance](#conformance)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This proposal extends instrumentation of tekton pipelines to include traces for pod events. Currently only reconciler related events are recorded as traces. The proposal is to capture container state changes (Started -> ContainerCreating -> Running -> Finished ) for each step of each task in the trace.
+
+## Motivation
+
+When we debug the performance of tekton pipelines, it is important to understand how much time is being spent on each step of the tasks. The total time of execution for a pipeline includes the time taken to initialize the pod, pull the images, run of init containers, sidecars etc. until completion. By tracing these phases, the end-users can easily visualize the steps in Jaeger UI.
+
+### Goals
+
+- Capture traces during phase change of a task pod
+  - Creation of container
+  - Running the container
+  - Finish running the container
+- Capture error and exit codes if the task was not successful due to failure in a step
+
+
+### Non-Goals
+
+- Tracing individual lines of each step
+- Tracing sidecars
+
+### Use Cases
+
+As a pipeline user:
+- I would like to understand the time taken by each step in a task so that I can optimize for performance
+- I would like to visualize the task steps in Jaeger to get a better understanding of the flow
+
+### Requirements
+
+- Enable tracing and provide Jaeger endpoint URL to the operator
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
```

Comment:
NIT: Could you include a short paragraph with a high-level description of the design from "## Design Details" in this section?